internal/graph/check.go

```diff
@@ -141,313 +139,226 @@ type CheckHandlerFunc func(ctx context.Context) (*ResolveCheckResponse, error)
+// to an allowed outcome. The first falsey causes premature termination of the reducer. Errors are swallowed if there is a false outcome.
 func intersection(ctx context.Context, concurrencyLimit int, handlers ...CheckHandlerFunc) (resp *ResolveCheckResponse, err error) {
-	if len(handlers) == 0 {
-		return &ResolveCheckResponse{
-			Allowed: false,
-		}, nil
+	if len(handlers) < 2 {
+		return nil, fmt.Errorf("%w, expected at least two rewrite operands for intersection operator, but got '%d'", openfgaErrors.ErrUnknown, len(handlers))
 	}
 
-	span := trace.SpanFromContext(ctx)
+	cancellableCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	ctx, cancel := context.WithCancel(ctx)
-	resultChan := make(chan checkOutcome, len(handlers))
+	pool := concurrency.NewPool(cancellableCtx, concurrencyLimit)
+	out := make(chan checkOutcome, len(handlers))
 
-	drain := resolver(ctx, concurrencyLimit, resultChan, handlers...)
+	for _, handler := range handlers {
+		h := handler // Capture loop variable for the goroutine
+		pool.Go(func(ctx context.Context) error {
+			concurrency.TrySendThroughChannel(cancellableCtx, runHandler(ctx, h), out)
+			return nil
+		})
+	}
 
-	defer func() {
-		cancel()
-		drainErr := drain()
-		if drainErr != nil {
-			err = drainErr
-		}
-		close(resultChan)
+	go func() {
+		_ = pool.Wait()
+		close(out)
 	}()
 
-	var elErr error
+	var finalErr error
+	finalResult := &ResolveCheckResponse{
+		Allowed: true,
+	}
+
 	for i := 0; i < len(handlers); i++ {
 		select {
-		case result := <-resultChan:
-			if result.err != nil {
-				telemetry.TraceError(span, result.err)
-				elErr = errors.Join(elErr, result.err)
+		case <-ctx.Done():
+			return nil, ctx.Err()
+
+		case outcome, ok := <-out:
+			if !ok {
+				break
+			}
+
+			if outcome.err != nil {
+				// Store the first error we see, but don't exit yet.
+				// A definitive 'false' result from another handler can override this.
+				if finalErr == nil {
+					finalErr = outcome.err
+				}
 				continue
 			}
 
-			if result.resp.GetCycleDetected() || !result.resp.GetAllowed() {
-				resp = result.resp
-				return
+			if outcome.resp.GetResolutionMetadata().CycleDetected || !outcome.resp.Allowed {
+				// Short-circuit failure. defer cancel() will clean up workers.
+				finalResult.Allowed = false
+				finalResult.ResolutionMetadata.CycleDetected = outcome.resp.GetResolutionMetadata().CycleDetected
+				return finalResult, nil
 			}
-		case <-ctx.Done():
-			err = ctx.Err()
-			return
 		}
 	}
 
-	// all operands are either truthy or we've seen at least one error
-	if elErr != nil {
-		err = elErr
-		return
+	// If we've processed all handlers without a definitive 'false',
+	// then any error we encountered along the way is the final result.
+	if finalErr != nil {
+		return nil, finalErr
 	}
 
-	resp = &ResolveCheckResponse{
-		Allowed: true,
-	}
-
-	return
+	// If the loop completes without any "false" outcomes or errors, the result is "true".
```

Comment:
I know this is not a change in functionality, do we want to call out anywhere as comments we do not propagate metadata such as datastore query count?

Reply:
I'm confused, what does this have to do in this context? DSQ has _long_ been gone from these reducers.